### PR TITLE
add path specificity scoring

### PR DESCRIPTION
**Pull request by kne42**
# Description
Score an fnmatch pattern, with higher specificities having lower scores.

Absolute paths have highest specificity, followed by paths with the most nesting, then by path segments with the least ambiguity.

replaces #4802

see #4613

### CODE REVIEW

**Review by DragaDoncila — DISMISSED** [review]
Love this, @kne42 ❤️. As mentioned I think we can remove `get_preferred_readers` if we're not using it anywhere yet, we can add it back if we need it. And as discussed, I think this actually doesn't need any UI changes, because the user should always be getting the reader they expect.

**Review by DragaDoncila — COMMENTED** [review]
Oops spoke too soon - few tests failing.

**Review by DragaDoncila — APPROVED** [review]
Looks great. I'm very happy for this to go in, I think it's a much more predictable system from the user's perspective.

File: napari/plugins/utils.py [hunk 1]
```diff
@@ -9,14 +12,104 @@
...
+
+    def add(match_flag):
+        score[-1] |= match_flag
+
+    # built-in fnmatch does not allow you to escape meta-characters
+    # so we don't need to handle them :)
+    for segment in segments:
+        # collapse foo/*/*/*.bar or foo*/*.bar but not foo*bar/*.baz
+        if segment and not (ends_with_star and segment.startswith('*')):
+            score.append(MatchFlag.NONE)
```
Comment (tlambert03) [inline napari/plugins/utils.py:58]: rather than the `add` function that modifies the last item here, maybe just use `continue` if you want to collapse, then start with `MatchFlag.None` and append it after all of the binary ops?

File: napari/plugins/utils.py [hunk 2]
```diff
@@ -9,14 +12,104 @@
...
+
+        if '*' in segment:
+            add(MatchFlag.STAR)
+        if '?' in segment:
+            add(MatchFlag.ANY)
+        if '[' in segment and ']' in segment[segment.index('[') :]:
+            add(MatchFlag.SET)
+
+        ends_with_star = segment != '' and segment[-1] == '*'
```
Comment (tlambert03) [inline napari/plugins/utils.py]:
```suggestion
        ends_with_star = segment.endswith('*')
```

?

File: napari/plugins/_tests/test_utils.py [hunk 3]
```diff
@@ -23,11 +25,170 @@ def test_get_preferred_reader_for_extension():
...
+        '/asdf/my-specific-folder/nested/my_file.tif'
+    )
+    assert reader == 'most-specific-plugin'
+
+
+def test_score_specificity_simple():
+    relpath, nested, score = score_specificity('')
+    assert relpath is True
+    assert nested == 0
+    assert score == [MatchFlag.NONE]
```
Comment (tlambert03) [inline napari/plugins/_tests/test_utils.py]: this is probably just subjective personal pref, so feel free to just ignore :)

I suppose you're going for self-documentation/readability here.  but also thing theres a lot to be said for the ease of visually parsing 6 repetitive lines vs 30 (it can be read more like a table with columns):

```python
assert score_specificity('') == (True, 0, [MatchFlag.NONE]) 
assert score_specificity('a') == (True, 0, [MatchFlag.NONE]) 
...
```

(could also consider a named tuple here if you like the readability it provides)

**Review by tlambert03 — APPROVED** [review]
this is really awesome.  love it.  very minor comments that you can feel free to ignore

File: napari/plugins/utils.py (hunk 2 above)
Comment (kne42) [reply napari/plugins/utils.py]: ahhh i was going to use that but i erroneously thought it belonged to a newer version of python... double checking now i got it confused with `removeprefix` ahaha 💀

File: napari/plugins/_tests/test_utils.py (hunk 3 above)
Comment (kne42) [reply napari/plugins/_tests/test_utils.py]: yes, good point, let me verify that black doesn't accidentally make it unreadable and if it looks good I'll change it to tuple comparisons

File: napari/plugins/utils.py (hunk 1 above)
Comment (kne42) [reply napari/plugins/utils.py:58]: tried this
```python
for segment in segments:
        if not segment or ends_with_star and segment.startswith('*'):
            # collapse foo/*/*/*.bar or foo*/*.bar but not foo*bar/*.baz
            continue
        
        segment_score = MatchFlag.NONE

        if '*' in segment:
            segment_score |= MatchFlag.STAR
        if '?' in segment:
            segment_score |= MatchFlag.ANY
        if '[' in segment and ']' in segment[segment.index('[') :]:
            segment_score |= MatchFlag.SET

        score.append(segment_score)

        ends_with_star = segment.endswith('*')
```

but realized that it actually failed an edge case: on `*/*?.tif` it will rate it as `Match.STAR` when it should actually collapse to `MatchFlag.STAR | MatchFlag.ANY`. capturing this in the tests now

Comment (tlambert03) [reply napari/plugins/utils.py:58]: ok!

Comment (kne42) [reply napari/plugins/utils.py:58]: maybe that was a bad example since `*?` is equal to `*`, a better example would be `*/*.t?f`

**Comment by kne42** [comment]
I've addressed all requested changes, please let me know if you have any other suggestions, otherwise I will merge this in 24 hours.

File: napari/plugins/_tests/test_utils.py (hunk 3 above)
Comment (tlambert03) [reply napari/plugins/_tests/test_utils.py]: (and if it does for some reason, I wouldn't hesitate to use `# fmt: off` here)

**Comment by tlambert03** [comment]
looks great to me, thanks @kne42
